File: comp-org/homework-assignments/adder/adder-student.py
"""
Implementation of a 4-bit ripple-carry adder using full adders
"built" from half adders (with tests and a few conveniences).

EDCI 5004 Computer Organization for Educators
Clayton Cafiero <cbcafier@uvm.edu>
2025-09-21
"""

def bits_to_bin(bits):
    """
    Utility function for converting a list of bits
    into a number. Bits should be in ascending order
    of powers of two. This function is based on one
    of the solutions given in the previous homework.
    """
    b = 0
    for bit in bits:
        b = b * 2 + bit
    return b


def half_adder(a, b):
    """
    This is a *half adder*, which takes two single-bit operands
    and returns their sum and carry.
    """
    assert a in [0, 1]  # ensure arguments are 1-bit
    assert b in [0, 1]
    
    s = (a and not b) or (not a and b)   # XOR
    c = a and b  # AND

    return int(s), int(c)


def full_adder(a, b, c_in):
    """
    Here you will implement a *full adder*, constructed from
    two half adders and an OR gate.
    """
    assert a in [0, 1]  # ensure arguments are 1-bit
    assert b in [0, 1]

    # COMPLETE THIS FUNCTION
    # You should call half_adder() TWICE

    return int(s), int(c)


def ripple_carry_adder(a, b):
    """
    Here you will implement a 4-bit ripple-carry adder.
    This takes as arguments two 4-bit (unsigned) binary
    numbers. This should return a tuple the first element
    of which is the carry out, and the second element of
    which is a list of bits in descending powers of two.
    """ 
    assert 15 >= a >= 0      # ensure arguments are 4-bit
    assert 15 >= b >= 0

    a_0 = (a >> 0) & 1       # isolate individual bits of a
    a_1 = (a >> 1) & 1
    a_2 = (a >> 2) & 1
    a_3 = # COMPLETE THIS
    
    b_0 = (b >> 0) & 1      # isolate individual bits of b
    b_1 = # COMPLETE THIS
    b_2 = # COMPLETE THIS
    b_3 = # COMPLETE THIS

    # MAKE FOUR CALLS TO full_adder() HERE
    s_0, c = # COMPLETE THIS
    s_1, c = # COMPLETE THIS
    s_2, c = # COMPLETE THIS
    s_3, c = # COMPLETE THIS

    # descending powers of two
    return c, [s_3, s_2, s_1, s_0]


def plus(a, b):
    """
    Just a little demo wrapper for the ripple-carry adder.
    You can pass decimal or binary arguments, but they must
    be numeric and not strings.
    """
    carry, bits = ripple_carry_adder(a, b)
    sum_ = bits_to_bin(bits)
    print(f"Decimal: {a} + {b} = {sum_} carry {carry}.")
    print(f"Binary: {a:04b} + {b:04b} = {sum_:04b} carry {carry}.")
    print()      
    

if __name__ == '__main__':

    # DON'T MODIFY ANY OF THE CODE BELOW

    # test against truth table for half adder
    assert half_adder(0, 0) == (0, 0)
    assert half_adder(0, 1) == (1, 0)
    assert half_adder(1, 0) == (1, 0)
    assert half_adder(1, 1) == (0, 1)

    # test against truth table for full adder
    assert full_adder(0, 0, 0) == (0, 0)
    assert full_adder(0, 0, 1) == (1, 0)
    assert full_adder(0, 1, 0) == (1, 0)
    assert full_adder(0, 1, 1) == (0, 1)
    assert full_adder(1, 0, 0) == (1, 0)
    assert full_adder(1, 0, 1) == (0, 1)
    assert full_adder(1, 1, 0) == (0, 1)
    assert full_adder(1, 1, 1) == (1, 1)

    # tests for ripple-carry adder (without carry out / overflow)
    assert ripple_carry_adder(0b0000, 0b0000) == (0, [0, 0, 0, 0])  # 0 + 0 = 0, carry 0
    assert ripple_carry_adder(0b0001, 0b0000) == (0, [0, 0, 0, 1])  # 1 + 0 = 1, carry 0
    assert ripple_carry_adder(0b0000, 0b0001) == (0, [0, 0, 0, 1])  # 0 + 1 = 1, carry 0
    assert ripple_carry_adder(0b0001, 0b0001) == (0, [0, 0, 1, 0])  # 1 + 1 = 2, carry 0
    assert ripple_carry_adder(0b0010, 0b0010) == (0, [0, 1, 0, 0])  # 2 + 2 = 4, carry 0
    assert ripple_carry_adder(0b0100, 0b0100) == (0, [1, 0, 0, 0])  # 4 + 4 = 8, carry 0
    assert ripple_carry_adder(0b0011, 0b0011) == (0, [0, 1, 1, 0])  # 3 + 3 = 6, carry 0
    assert ripple_carry_adder(0b0101, 0b0101) == (0, [1, 0, 1, 0])  # 5 + 5 = 10, carry 0
    assert ripple_carry_adder(0b0111, 0b0111) == (0, [1, 1, 1, 0])  # 7 + 7 = 14, carry 0
    assert ripple_carry_adder(0b1000, 0b0111) == (0, [1, 1, 1, 1])  # 8 + 7 = 15, carry 0
    assert ripple_carry_adder(0b1010, 0b0101) == (0, [1, 1, 1, 1])  # 10 + 5 = 15, carry 0
    assert ripple_carry_adder(0b0110, 0b0110) == (0, [1, 1, 0, 0])  # 6 + 6 = 12, carry 0

    # tests for ripple-carry adder (with carry out / overflow)
    # The carry bit represents the 2^4 (sixteens) bit
    assert ripple_carry_adder(0b1111, 0b0001) == (1, [0, 0, 0, 0])  # 15 + 1 = 0, carry 1
    assert ripple_carry_adder(0b1110, 0b1110) == (1, [1, 1, 0, 0])  # 14 + 14 = 12, carry 1 
    assert ripple_carry_adder(0b0111, 0b1100) == (1, [0, 0, 1, 1])  # 7 + 12 = 3, carry 1
    assert ripple_carry_adder(0b1001, 0b1001) == (1, [0, 0, 1, 0])  # 9 + 9 = 2, carry 1
    assert ripple_carry_adder(0b1111, 0b1111) == (1, [1, 1, 1, 0])  # 15 + 15 = 14, carry 1
    
    # If all tests passed, give a little demo
    plus(0, 0)
    plus(1, 1)
    plus(3, 5)
    plus(7, 7)
    plus(10, 10)
    plus(9, 6)
    plus(12, 9)
    plus(15, 15)

File: comp-org/homework-assignments/adder/assignment/adder-student.c
// adder.c
// Build:  gcc -std=c11 -Wall -Wextra adder.c -o adder
// Run:    ./adder

#include <stdio.h>
#include <assert.h>

typedef struct { int sum; int carry; } SumCarry;      // for half/full adder
typedef struct { int bits[4]; int carry;} RCAResult; // for ripple-carry adder

// Utilities
int bits_to_bin(const int *bits, int len) {
/*
 * Utility function for converting a list of bits 
 * into a number. Bits should be in ascending order 
 * of powers of two. This function is based on one 
 * of the solutions given in the previous homework.
 */
    int b = 0;
    for (int i = 0; i < len; ++i) {
        b = b * 2 + bits[i];   // same as (b << 1) | bits[i]
    }
    return b;
}

// Helper for printing binary
void print_bin4(int x) {
    for (int i = 3; i >= 0; --i) {
        putchar(((x >> i) & 1) ? '1' : '0');
    }
}

// half_adder, full_adder return `SumCarry struct to mimic returning tuple in python -> Operations relatively the same
SumCarry half_adder(int a, int b) {
  /*
   *This is a *half adder*, which takes two single-bit operands and returns their sum and carry.
   */
    assert(a == 0 || a == 1);
    assert(b == 0 || b == 1);

    SumCarry r;
    r.sum   = (a ^ b);   // XOR
    r.carry = (a & b);   // AND
    return r;
}

SumCarry full_adder(int a, int b, int c_in) {
  /*
   * Here you will implement a *full adder*, constructed from two half adders and an OR gate.
  */
    assert(a == 0 || a == 1);
    assert(b == 0 || b == 1);
    assert(c_in == 0 || c_in == 1);
    
    // COMPLETE THIS FUNCTION, call half_adder TWICE
    
    return r;
}

// ripple_carry_adder returns RCAResult struct with final int in binary plus carry if applicable
RCAResult ripple_carry_adder(unsigned a, unsigned b) {
  /*
   * Here you will implement a 4-bit ripple-carry adder.
   * This takes as arguments two 4-bit (unsigned) binary 
   * numbers. This should return a tuple the first element 
   * of which is the carry out, and the second element of 
   * which is a list of bits in descending powers of two.
  */
    assert(a <= 15); // 4-bit
    assert(b <= 15);

    int a0 = (a >> 0) & 1;
    int a1 = (a >> 1) & 1;
    int a2 = (a >> 2) & 1;
    int a3 = // COMPLETE

    int b0 = (b >> 0) & 1;
    int b1 = // COMPLETE 
    int b2 = // COMPLETE 
    int b3 = // COMPLETE 

    // MAKE 4 CALLS TO full_adder
    SumCarry s0 = // COMPLETE 
    SumCarry s1 = // COMPLETE 
    SumCarry s2 = // COMPLETE 
    SumCarry s3 = // COMPLETE 

    RCAResult r;
    r.bits[0] = s3.sum;  // MSB
    r.bits[1] = s2.sum;
    r.bits[2] = s1.sum;
    r.bits[3] = s0.sum;  // LSB
    r.carry   = s3.carry;

    return r;
}

// DO NOT MODIFY ANY CODE BELOW

// Plus wrapper 
void plus(unsigned a, unsigned b) {
    RCAResult R = ripple_carry_adder(a, b);
    int sum_ = bits_to_bin(R.bits, 4);

    printf("Decimal: %u + %u = %d carry %d.\n", a, b, sum_, R.carry);
    printf("Binary : ");
    print_bin4(a);
    printf(" + ");
    print_bin4(b);
    printf(" = ");
    print_bin4(sum_);
    printf(" carry %d.\n\n", R.carry);
}

// Helper functions for asserting test cases similarly to python, not sure if there is a test framework for C that I should be using
void assertion_sumcarry(SumCarry got, int exp_sum, int exp_carry) {
    assert(got.sum   == exp_sum);
    assert(got.carry == exp_carry);
}

void assertion_RCA(unsigned a, unsigned b, int exp_carry,
                      int e3, int e2, int e1, int e0) {
    RCAResult r = ripple_carry_adder(a, b);
    assert(r.carry   == exp_carry);
    assert(r.bits[0] == e3);
    assert(r.bits[1] == e2);
    assert(r.bits[2] == e1);
    assert(r.bits[3] == e0);
}

int main(void) {
    // Half adder cases
    assertion_sumcarry(half_adder(0,0), 0,0);
    assertion_sumcarry(half_adder(0,1), 1,0);
    assertion_sumcarry(half_adder(1,0), 1,0);
    assertion_sumcarry(half_adder(1,1), 0,1);

    // Full adder cases
    assertion_sumcarry(full_adder(0,0,0), 0,0);
    assertion_sumcarry(full_adder(0,0,1), 1,0);
    assertion_sumcarry(full_adder(0,1,0), 1,0);
    assertion_sumcarry(full_adder(0,1,1), 0,1);
    assertion_sumcarry(full_adder(1,0,0), 1,0);
    assertion_sumcarry(full_adder(1,0,1), 0,1);
    assertion_sumcarry(full_adder(1,1,0), 0,1);
    assertion_sumcarry(full_adder(1,1,1), 1,1);

    // Ripple carry cases
    assertion_RCA(0,  0,  0, 0,0,0,0);  // 0000 + 0000 = 0000
    assertion_RCA(1,  0,  0, 0,0,0,1);  // 0001 + 0000 = 0001
    assertion_RCA(0,  1,  0, 0,0,0,1);  // 0000 + 0001 = 0001
    assertion_RCA(1,  1,  0, 0,0,1,0);  // 0001 + 0001 = 0010
    assertion_RCA(2,  2,  0, 0,1,0,0);  // 0010 + 0010 = 0100
    assertion_RCA(4,  4,  0, 1,0,0,0);  // 0100 + 0100 = 1000
    assertion_RCA(3,  3,  0, 0,1,1,0);  // 0011 + 0011 = 0110
    assertion_RCA(5,  5,  0, 1,0,1,0);  // 0101 + 0101 = 1010
    assertion_RCA(7,  7,  0, 1,1,1,0);  // 0111 + 0111 = 1110
    assertion_RCA(8,  7,  0, 1,1,1,1);  // 1000 + 0111 = 1111
    assertion_RCA(10, 5,  0, 1,1,1,1);  // 1010 + 0101 = 1111
    assertion_RCA(6,  6,  0, 1,1,0,0);  // 0110 + 0110 = 1100
    assertion_RCA(15, 1,  1, 0,0,0,0);  // 1111 + 0001 = 0000, carry 1
    assertion_RCA(14,14,  1, 1,1,0,0);  // 1110 + 1110 = 1100, carry 1
    assertion_RCA(7, 12,  1, 0,0,1,1);  // 0111 + 1100 = 0011, carry 1
    assertion_RCA(9,  9,  1, 0,0,1,0);  // 1001 + 1001 = 0010, carry 1
    assertion_RCA(15,15,  1, 1,1,1,0);  // 1111 + 1111 = 1110, carry 1

    // Plus demo if cases pass
    plus(0, 0);
    plus(1, 1);
    plus(3, 5);
    plus(7, 7);
    plus(10, 10);
    plus(9, 6);
    plus(12, 9);
    plus(15, 15);

    puts("All tests passed.");
    return 0;
}
